In [36]:
from sklearn.ensemble import StackingClassifier, VotingClassifier, RandomForestClassifier, GradientBoostingClassifier
from sklearn.datasets import make_classification
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, recall_score, precision_score,confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler

In [3]:
df = pd.read_csv("novagen_dataset.csv")

In [4]:
df.head()

,Age,BMI,Blood_Pressure,Cholesterol,Glucose_Level,Heart_Rate,Sleep_Hours,Exercise_Hours,Water_Intake,Stress_Level,...,Diet,MentalHealth,PhysicalActivity,MedicalHistory,Allergies,Diet_Type__Vegan,Diet_Type__Vegetarian,Blood_Group_AB,Blood_Group_B,Blood_Group_O
0,2.0,26.0,111.0,198.0,99.0,72.0,4.0,1.0,5.0,5.0,...,1,2,1,0,1,False,True,True,False,False
1,8.0,24.0,121.0,199.0,103.0,75.0,2.0,1.0,2.0,9.0,...,1,2,1,2,2,False,False,True,False,False
2,81.0,27.0,147.0,203.0,100.0,74.0,10.0,-0.0,5.0,1.0,...,2,0,0,1,0,True,False,False,False,False
3,25.0,21.0,150.0,199.0,102.0,70.0,7.0,3.0,3.0,3.0,...,1,2,1,2,0,True,False,False,True,False
4,24.0,26.0,146.0,202.0,99.0,76.0,10.0,2.0,5.0,1.0,...,2,0,2,0,2,False,True,False,True,False


In [5]:
df.isnull().sum()

Age                      0
BMI                      0
Blood_Pressure           0
Cholesterol              0
Glucose_Level            0
Heart_Rate               0
Sleep_Hours              0
Exercise_Hours           0
Water_Intake             0
Stress_Level             0
Target                   0
Smoking                  0
Alcohol                  0
Diet                     0
MentalHealth             0
PhysicalActivity         0
MedicalHistory           0
Allergies                0
Diet_Type__Vegan         0
Diet_Type__Vegetarian    0
Blood_Group_AB           0
Blood_Group_B            0
Blood_Group_O            0
dtype: int64

In [7]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
9544    False
9545    False
9546    False
9547    False
9548    False
Length: 9549, dtype: bool

In [12]:
x = df.drop(columns = ["Target"])
y = df["Target"]

In [14]:
x_train,x_test,y_train,y_test = train_test_split(
    x,
    y,
    test_size = 0.2,
    random_state = 40
)

In [15]:
ss = StandardScaler()

In [17]:
x_train_scaled = ss.fit_transform(x_train)
x_test_scaled = ss.transform(x_test)


LOgistic Regression

In [37]:
lr = LogisticRegression(
    penalty = "l2",
    solver = "liblinear",
    max_iter = 1000
)
lr.fit(x_train_scaled,y_train)
y_pred_lr = lr.predict(x_test_scaled)
print("Logistic Regression Accuracy:", accuracy_score(y_test, y_pred_lr))
print("Logistic Regression Recall:", recall_score(y_test, y_pred_lr))
print("Logistic Regression precision:", precision_score(y_test, y_pred_lr))
print("Logistic Regression confusion matrix:", confusion_matrix(y_test, y_pred_lr))
print(classification_report(y_test, y_pred_lr))


Logistic Regression Accuracy: 0.805759162303665
Logistic Regression Recall: 0.8206896551724138
Logistic Regression precision: 0.815068493150685
Logistic Regression confusion matrix: [[706 189]
 [182 833]]
              precision    recall  f1-score   support

           0       0.80      0.79      0.79       895
           1       0.82      0.82      0.82      1015

    accuracy                           0.81      1910
   macro avg       0.81      0.80      0.80      1910
weighted avg       0.81      0.81      0.81      1910



# knn

In [38]:
knn = KNeighborsClassifier(
     n_neighbors=5,
    metric="euclidean"
    
)
knn.fit(x_train_scaled,y_train)
y_pred_knn = knn.predict(x_test_scaled)
print("KNN Accuracy:", accuracy_score(y_test, y_pred_knn))
print("KNN Recall:", recall_score(y_test, y_pred_knn))
print("KNN precision:", precision_score(y_test, y_pred_knn))
print("KNN confusion matrix:", confusion_matrix(y_test, y_pred_knn))
print(classification_report(y_test, y_pred_knn))


KNN Accuracy: 0.8759162303664921
KNN Recall: 0.870935960591133
KNN precision: 0.8929292929292929
KNN confusion matrix: [[789 106]
 [131 884]]
              precision    recall  f1-score   support

           0       0.86      0.88      0.87       895
           1       0.89      0.87      0.88      1015

    accuracy                           0.88      1910
   macro avg       0.88      0.88      0.88      1910
weighted avg       0.88      0.88      0.88      1910



# Naive bayes

In [39]:
NB = GaussianNB()
NB.fit(x_train_scaled,y_train)
y_pred_NB = NB.predict(x_test_scaled)
print("NB Accuracy:", accuracy_score(y_test, y_pred_NB))
print("NB Recall:", recall_score(y_test, y_pred_NB))
print("NB precision:", precision_score(y_test, y_pred_NB))
print("NB confusion matrix:", confusion_matrix(y_test, y_pred_NB))
print(classification_report(y_test, y_pred_NB))


NB Accuracy: 0.8005235602094241
NB Recall: 0.7842364532019704
NB precision: 0.8308977035490606
NB confusion matrix: [[733 162]
 [219 796]]
              precision    recall  f1-score   support

           0       0.77      0.82      0.79       895
           1       0.83      0.78      0.81      1015

    accuracy                           0.80      1910
   macro avg       0.80      0.80      0.80      1910
weighted avg       0.80      0.80      0.80      1910



Decision tree

In [44]:
dt = DecisionTreeClassifier(
     max_depth = 4,
    min_samples_leaf = 4,
    min_samples_split = 4,
    random_state = 42
)
dt.fit(x_train_scaled,y_train)
y_pred_dt = dt.predict(x_test_scaled)
print("Decision Tree Accuracy:", accuracy_score(y_test, y_pred_dt))
print("Decision Tree Recall:", recall_score(y_test, y_pred_dt))
print("Decision Tree:", precision_score(y_test, y_pred_dt))
print("Decision Tree matrix:", confusion_matrix(y_test, y_pred_dt))
print(classification_report(y_test, y_pred_dt))


Decision Tree Accuracy: 0.8251308900523561
Decision Tree Recall: 0.8748768472906404
Decision Tree: 0.810958904109589
Decision Tree matrix: [[688 207]
 [127 888]]
              precision    recall  f1-score   support

           0       0.84      0.77      0.80       895
           1       0.81      0.87      0.84      1015

    accuracy                           0.83      1910
   macro avg       0.83      0.82      0.82      1910
weighted avg       0.83      0.83      0.82      1910



# Random forest

In [45]:
rf = RandomForestClassifier(
    n_estimators = 201,
    max_depth = 4,
    min_samples_leaf = 4,
    min_samples_split = 4,
    ccp_alpha = 0.01,
    random_state = 42
)
rf.fit(x_train_scaled,y_train)
y_pred_rf = rf.predict(x_test_scaled)
print("Random forest Accuracy:", accuracy_score(y_test, y_pred_rf))
print("Random forest Recall:", recall_score(y_test, y_pred_rf))
print("Random forest:", precision_score(y_test, y_pred_rf))
print("Random forest matrix:", confusion_matrix(y_test, y_pred_rf))
print(classification_report(y_test, y_pred_rf))


Random forest Accuracy: 0.8068062827225131
Random forest Recall: 0.8817733990147784
Random forest: 0.7823426573426573
Random forest matrix: [[646 249]
 [120 895]]
              precision    recall  f1-score   support

           0       0.84      0.72      0.78       895
           1       0.78      0.88      0.83      1015

    accuracy                           0.81      1910
   macro avg       0.81      0.80      0.80      1910
weighted avg       0.81      0.81      0.81      1910



# Gradient boosting

In [46]:
gb = RandomForestClassifier(
    n_estimators = 201,
    max_depth = 4,
    min_samples_leaf = 4,
    min_samples_split = 4,
    ccp_alpha = 0.01,
    random_state = 42
)
gb.fit(x_train_scaled,y_train)
y_pred_gb = gb.predict(x_test_scaled)
print("Gradient boosting Accuracy:", accuracy_score(y_test, y_pred_gb))
print("Gradient boosting Recall:", recall_score(y_test, y_pred_gb))
print("Gradient boosting:", precision_score(y_test, y_pred_gb))
print("Gradient boosting matrix:", confusion_matrix(y_test, y_pred_gb))
print(classification_report(y_test, y_pred_gb))

Gradient boosting Accuracy: 0.8068062827225131
Gradient boosting Recall: 0.8817733990147784
Gradient boosting: 0.7823426573426573
Gradient boosting matrix: [[646 249]
 [120 895]]
              precision    recall  f1-score   support

           0       0.84      0.72      0.78       895
           1       0.78      0.88      0.83      1015

    accuracy                           0.81      1910
   macro avg       0.81      0.80      0.80      1910
weighted avg       0.81      0.81      0.81      1910



# Voting Classifier

In [48]:
vc = VotingClassifier(
    estimators = [
        ("lr", lr),
        ("dt", dt),
        ("knn", knn),
        ("NB", NB)
    ],
    voting = "soft"
)
vc.fit(x_train_scaled,y_train)
y_pred_vc = vc.predict(x_test_scaled)
print("Voting Classifier Accuracy:", accuracy_score(y_test, y_pred_vc))
print("Voting Classifier Recall:", recall_score(y_test, y_pred_vc))
print("Voting Classifier:", precision_score(y_test, y_pred_vc))
print("Voting Classifier matrix:", confusion_matrix(y_test, y_pred_vc))
print(classification_report(y_test, y_pred_vc))

Voting Classifier Accuracy: 0.8612565445026178
Voting Classifier Recall: 0.8857142857142857
Voting Classifier: 0.857824427480916
Voting Classifier matrix: [[746 149]
 [116 899]]
              precision    recall  f1-score   support

           0       0.87      0.83      0.85       895
           1       0.86      0.89      0.87      1015

    accuracy                           0.86      1910
   macro avg       0.86      0.86      0.86      1910
weighted avg       0.86      0.86      0.86      1910



# Stacking classifier

In [50]:
sc = StackingClassifier(
    estimators = [
        ("lr", lr),
        ("dt", dt),
        ("knn", knn),
        ("NB", NB)
    ],
    final_estimator = dt,
    cv = 5
)
sc.fit(x_train_scaled,y_train)
y_pred_sc = sc.predict(x_test_scaled)
print("Stacking Classifier Accuracy:", accuracy_score(y_test, y_pred_sc))
print("Stacking Classifier Recall:", recall_score(y_test, y_pred_sc))
print("Stacking Classifier:", precision_score(y_test, y_pred_sc))
print("Stacking Classifier matrix:", confusion_matrix(y_test, y_pred_sc))
print(classification_report(y_test, y_pred_sc))

Stacking Classifier Accuracy: 0.8858638743455497
Stacking Classifier Recall: 0.8837438423645321
Stacking Classifier: 0.8996990972918756
Stacking Classifier matrix: [[795 100]
 [118 897]]
              precision    recall  f1-score   support

           0       0.87      0.89      0.88       895
           1       0.90      0.88      0.89      1015

    accuracy                           0.89      1910
   macro avg       0.89      0.89      0.89      1910
weighted avg       0.89      0.89      0.89      1910



## Results

| Model                | Recall |
|----------------------|:------:|
| Logistic Regression  | 82.06%  |
| KNN                  | 87.09%  |
| Naive Bayes          | 78.42%  |
| Decision Tree        | 87.48%  |
| Random Forest        | 88.17%  |
| Gradient Boosting    | 88.17%  |
| Voting Classifier    | 88.57%  |
| Stacking Classifier  | 88.37%  |


### Best Classifier that we should use for NovaGen(based on Recall) - Stacking Classifier with accuracy of 88.58%